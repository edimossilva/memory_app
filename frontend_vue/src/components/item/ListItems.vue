<template>
  <div>
    <div class="mb-4">
      <b-field label="search" label-position="inside" type="is-primary">
        <b-input type="text" v-model="filter" data-cy="filter"></b-input>
      </b-field>
    </div>
    <div class="columns is-multiline is-mobile">
      <div class="column is-half" v-for="item in filteredItems" :key="item.key">
        <show-item :item="item"></show-item>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState, mapActions } from "vuex";
import ShowItem from "@/components/item/ShowItem.vue";
import { getItemsApi } from "../../services/itemsApi";

export default {
  name: "ListItems",
  mounted() {
    getItemsApi().then(
      (response) => {
        this.setItems(response.data.data);
      },
      (error) => {
        console.log(error);
      }
    );
  },
  data() {
    return {
      item: { key: "", value: "" },
      filter: "",
    };
  },
  components: { ShowItem },
  computed: {
    ...mapState(["items"]),
    filteredItems() {
      const { items, filter } = this;
      if (filter) return items.filter((item) => item.key.includes(filter));
      return items;
    },
  },
  methods: {
    ...mapActions(["setItems"]),
  },
};
</script>
