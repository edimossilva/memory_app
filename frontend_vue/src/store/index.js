import Vue from "vue";
import Vuex from "vuex";

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
    items: [],
  },
  mutations: {
    addItem(state, item) {
      state.items.push(item);
    },
    removeItem(state, item) {
      const index = state.items.findIndex((i) => i.key === item.key);
      state.items.splice(index, 1);
    },
    setItems(state, items) {
      state.items = items;
    },
    // mutate(state, payload) {
    //   state[payload.property] = payload.value;
    // },
  },
  actions: {
    addItem({ commit }, item) {
      commit("addItem", { ...item });
    },
    removeItem({ commit }, item) {
      commit("removeItem", item);
    },
    setItems({ commit }, items) {
      commit("setItems", items);
    },
  },
  modules: {},
});
