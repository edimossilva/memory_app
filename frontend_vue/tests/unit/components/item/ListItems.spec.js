import { shallowMount } from "@vue/test-utils";
import ListItems from "@/components/item/ListItems.vue";
import Vue from "vue";
import VueRouter from "vue-router";
import store from "@/store/index";

Vue.use(VueRouter);
const item1 = { key: "mockKey1", value: "mockValue1" };
const item2 = { key: "mockKey2", value: "mockValue2" };
const mockItems = [item1, item2];
describe("ListItems.vue", () => {
  it("Matchs Snapshot", () => {
    const wrapper = shallowMount(ListItems, {
      store,
      computed: { items: () => mockItems },
    });
    expect(wrapper).toMatchSnapshot();
  });
  describe("computed", () => {
    describe("filteredItems", () => {
      describe("When filter is empty", () => {
        const wrapper = shallowMount(ListItems, {
          store,
          propsData: { filter: "" },
          computed: { items: () => mockItems },
        });
        it("should has all items", () => {
          expect(wrapper.vm.filteredItems).toBe(mockItems);
        });
      });
      describe("When filter match on item", () => {
        const wrapper = shallowMount(ListItems, {
          store,
          computed: { items: () => mockItems },
        });

        wrapper.setData({
          filter: "Key1",
        });

        it("should has only filtered item", () => {
          expect(wrapper.vm.filteredItems).toStrictEqual([item1]);
        });
      });
    });
  });
});
