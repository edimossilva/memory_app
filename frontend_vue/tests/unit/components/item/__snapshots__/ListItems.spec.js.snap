// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItems.vue Matchs Snapshot 1`] = `
<div>
  <div class="mb-4">
    <b-field label="search" label-position="inside" type="is-primary">
      <b-input type="text" data-cy="filter"></b-input>
    </b-field>
  </div>
  <div class="columns is-multiline is-mobile">
    <div class="column is-half">
      <show-item-stub item="[object Object]"></show-item-stub>
    </div>
    <div class="column is-half">
      <show-item-stub item="[object Object]"></show-item-stub>
    </div>
  </div>
</div>
`;
